### ISSUE TITLE
[1.16.5] Badly Compressed Packet

### ISSUE THREAD

**Issue by TooCoolCello** [Teneted/Tenet#1998]
**Minecraft Version :** 1.16.5

**Mohist Version :** 919-922

**Operating System :** Windows 10 Pro

**Log :** https://gist.github.com/TooCoolCello/70a2e95e63348fe4395d5d448935dd43

**Mod list :** https://www.toptal.com/developers/hastebin/bugogodele.apache

**Plugin list :** None

**Description of issue :**  I submitted this issue previously and saw the github update that was supposed to fix it but still my server starts fine but upon a player joining they are kicked with the error:
"Internal Exception: io.netty.handler.codec.DecoderException: Badly compressed packet - size of 4058499 is larger than protocol maximum of 2097152"

I have used RandomPatches aswell as Connectivity to try and raise or disable the packet limit with no success.

**Comment by TooCoolCello** [comment]
Bump

**Comment by Mgazul** [comment]
Provide mod package for easy testing

**Comment by TooCoolCello** [comment]
Here is my mods folder
https://drive.google.com/file/d/1t2NeVdJxxfzPHcttosGNHIEIqEqgwCMN/view?usp=sharing

**Comment by Kizwiz** [comment]
This also occurs on the Mineshafts and Monsters modpack if that helps.

EDIT: Does mohist interfere with random patches mod? I believe that mod is meant to resolve the Badly Compressed Packets issue. Just a thought

**Comment by TooCoolCello** [comment]
> EDIT: Does mohist interfere with random patches mod? I believe that mod is meant to resolve the Badly Compressed Packets issue. Just a thought

I know that it somehow messes with the packet size after random patches does... They've implemented a couple fixes but nothings worked yet.

**Comment by TigerWalts** [comment]
The problem is also evident on Mohist 1.16.5 `958` while running Vault Hunters `1.12.4`.

TLDR: A custom registry added by CraftTweaker is being synced to the client in an uncompressed (or un-chunked) packet.

---

I was able to connect after adding the **Connectivity** mod to the client. This mod catches the errors and ignores them.

The mod has an option to log these packets too. I turned this on and I discovered:

- The packet being sent appears to contain data for CraftTweaker
- Some of that data looked like registry or indexed lookup data (or possibly bytecoded CraftTweaker api calls)
- Most of the data is the contents of the tweak script files
- The packet is uncompressed and the total bytes logged matches the amount that appears in the error message
- No unrelated data can be seen

It looks like the packet is not being compressed. If it was then it would be a fraction of the size.

From a quick look at the CraftTweaker code, it looks like this might be a packet to sync a custom registry called `crafttweaker:scripts`. I can't see any other way in the code that the mod could sync this data.

It's then left up to either Minecraft (or Forge) to sync the registry.

Under Forge the packet would be compressed (or chunked), but running with Mohist it isn't.

**Comment by ghost** [comment]
Maybe we should revert the change of PROTOCOL_MAX from 16777216 to the original 2097152 as in VanillaPacketSplitter I found that if the packet is smaller than PROTOCOL_MAX the packet would be directly sended to client without splitting.
`            if (buf.readableBytes() <= PROTOCOL_MAX)
            {
                buf.release();
                out.add(packet);
            }`

**Comment by Git-Ricardox** [comment]
Hey, i installed connectivity too..but it doesn't work for me...after removing randompatches it did! However i'am not able to connect to my Mohist Server!


> The problem is also evident on Mohist 1.16.5 `958` while running Vault Hunters `1.12.4`.
> 
> TLDR: A custom registry added by CraftTweaker is being synced to the client in an uncompressed (or un-chunked) packet.
> 
> I was able to connect after adding the **Connectivity** mod to the client. This mod catches the errors and ignores them.
> 
> The mod has an option to log these packets too. I turned this on and I discovered:
> 
> * The packet being sent appears to contain data for CraftTweaker
> * Some of that data looked like registry or indexed lookup data (or possibly bytecoded CraftTweaker api calls)
> * Most of the data is the contents of the tweak script files
> * The packet is uncompressed and the total bytes logged matches the amount that appears in the error message
> * No unrelated data can be seen
> 
> It looks like the packet is not being compressed. If it was then it would be a fraction of the size.
> 
> From a quick look at the CraftTweaker code, it looks like this might be a packet to sync a custom registry called `crafttweaker:scripts`. I can't see any other way in the code that the mod could sync this data.
> 
> It's then left up to either Minecraft (or Forge) to sync the registry.
> 
> Under Forge the packet would be compressed (or chunked), but running with Mohist it isn't.
